fibers/conditions.scm

```diff
@@ -80,12 +105,25 @@ returns @code{#t} otherwise."
        (#t ;; Cvar already signalled.
         #f)))))
 
+;; How many times we run the block-fn until we gc
+(define %steps-till-gc 42)  ; haven't tried testing for the most efficient number
+
 (define (wait-operation cvar)
   "Make an operation that will complete when @var{cvar} is signalled."
   (match cvar
-    (($ <condition> signalled? waiters)
+    (($ <condition> signalled? waiters gc-step)
      (define (try-fn) (and (atomic-box-ref signalled?) values))
      (define (block-fn flag sched resume)
+       ;; Increment the garbage collection counter.
+       ;; If we've surpassed the number of steps until garbage collection,
+       ;; prune out waiters that have already succeeded.
+       (when (> (counter-increment! gc-step)
```

Comment:
Here I think we should be precise: i.e. you don't want multiple threads which both increment the counter past the limit to be racing to clear the waiters list.  Consider also making the counter count *down* rather than up -- that way the condition checks it against zero rather than a constant, and resetting it chooses the initial value.